groovy-language-server/src/main/java/com/palantir/ls/util/SourceWriter.java

```diff
@@ -0,0 +1,167 @@
+        this.source = source;
+        this.destination = destination;
+    }
+
+    /**
+     * Creates a SourceWriter which creates a copy of source at destination.
+     */
+    public static SourceWriter of(Path source, Path destination) throws IOException {
+        SourceWriter writer = new SourceWriter(source, destination);
+        writer.initialize();
+        return writer;
+    }
+
+    public Path getSource() {
+        return source;
+    }
+
+    public Path getDestination() {
+        return destination;
+    }
+
+    /**
+     * Applies the given changes to the destination file. Does not handle intersecting ranges in the changes.
+     */
+    public synchronized void applyChanges(List<? extends TextDocumentContentChangeEvent> changes) throws IOException {
+        // From earliest start of range to latest
+        List<TextDocumentContentChangeEvent> sortedChanges =
+                changes.stream().sorted((c1, c2) -> Ranges.POSITION_COMPARATOR.compare(c1.getRange().getStart(),
+                        c2.getRange().getStart())).collect(Collectors.toList());
+
+        // TODO(#60): Check if any of the changes intersect and throw an exception
+
+        String newFileContents = handleChanges(sortedChanges);
+
+        FileUtils.writeStringToFile(destination.toFile(), newFileContents, StandardCharsets.UTF_8.toString());
+    }
+
+    private String handleChanges(List<TextDocumentContentChangeEvent> sortedChanges) throws IOException {
+        StringBuilder newFileContents = new StringBuilder();
+        BufferedReader file =
+                new BufferedReader(new InputStreamReader(new FileInputStream(destination.toAbsolutePath().toString()),
+                        StandardCharsets.UTF_8.toString()));
+        String line;
+        int lineNum = 0;
+        int sortedChangesIndex = 0;
+        while ((line = file.readLine()) != null) {
+            if (sortedChangesIndex >= sortedChanges.size()) {
+                newFileContents.append(line).append("\n");
+                continue;
+            }
+            Position start = sortedChanges.get(sortedChangesIndex).getRange().getStart();
+            Position end = sortedChanges.get(sortedChangesIndex).getRange().getEnd();
+
+            if (start.getLine() == lineNum) {
+                // Add part of line before our change
+                if (end.getCharacter() > 0) {
+                    // We need to include part of the line before our change
+                    newFileContents.append(line.substring(0, Math.min(start.getCharacter(), line.length())));
+                }
+                if (end.getLine() == lineNum) {
+                    // Single line range
+                    newFileContents.append(sortedChanges.get(sortedChangesIndex).getText());
+                    // Add part of line after our change
+                    if (end.getCharacter() + 1 < line.length()) {
+                        newFileContents.append(line.substring(Math.max(end.getCharacter() + 1, 0)));
```

Comment:
It could if they gave an negative character. I actually have a test that does this. The idea is that if they want to add at the start of a line.
I don't know if in practice it could happen. If I add an 'x' at the start of a line, does the whole line get sent as a change and the whole range of the line get replaced?